### PR TITLE
GEOMESA-708 Expose BatchWriterConfig

### PR DESCRIPTION
**Pull request by jnh5y**
* GeoMesaBatchWriterConfig looks for various settings on the classpath and uses them when creating BatchWriterConfigs.

Signed-off-by: Jim Hughes <jnh5y@ccri.com>

### CODE REVIEW

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.concurrent.TimeUnit
+
+import com.typesafe.scalalogging.slf4j.Logging
+import org.apache.accumulo.core.client.BatchWriterConfig
+
+import scala.util.Try
+
+object GeoMesaBatchWriterConfig extends Logging {
+  val WRITERLATENCYSECONDS = "geomesa.batchwriter.latency.seconds"  // Measured in seconds
```
Comment (jahhulbert-ccri) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: I think the convention in scala is to use UpperCamelCase for constants and in java its all upper with underscores...so perhaps ```WRITER_LATENCY_SECONDS```

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        bwc.setMaxLatency(latency, TimeUnit.SECONDS)
+      case None =>
+        val milliLatency = fetchProperty(WRITERLATENCYMILLIS).getOrElse(DEFAULTLATENCY)
+        logger.trace(s"GeoMesaBatchWriter config: maxLatency set to $milliLatency milliseconds.")
+        bwc.setMaxLatency(milliLatency, TimeUnit.MILLISECONDS)
+    }
+
+    val memory = fetchProperty(WRITERMEMORY).getOrElse(DEFAULTMAXMEMORY)
+    logger.trace(s"GeoMesaBatchWriter config: maxMemory set to $memory bytes.")
+    bwc.setMaxMemory(memory)
```
Comment (jahhulbert-ccri) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: @jnh5y does this work with the strings ```100M``` or ```300k```  ? If not can we open a ticket to support that in config similar to Java opts? I think there is some accumulo code or hadoop code somewhere we can nab.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala (hunk 1 above)
Comment (jnh5y) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: Underscores seems to be our preferred approach.  Changing.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala (hunk 2 above)
Comment (jnh5y) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: Added ticket https://geomesa.atlassian.net/browse/GEOMESA-735 and linked to it in a comment.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.typesafe.scalalogging.slf4j.Logging
+import org.apache.accumulo.core.client.BatchWriterConfig
+
+import scala.util.Try
+
+object GeoMesaBatchWriterConfig extends Logging {
+  val WRITER_LATENCY_SECONDS = "geomesa.batchwriter.latency.seconds"  // Measured in seconds
+  val WRITER_LATENCY_MILLIS  = "geomesa.batchwriter.latency.millis"   // Measured in millis
+  val WRITER_MEMORY         = "geomesa.batchwriter.memory"           // Measured in bytes
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: [picky] line up the =

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    // https://geomesa.atlassian.net/browse/GEOMESA-735
+    val memory = fetchProperty(WRITER_MEMORY).getOrElse(DEFAULT_MAX_MEMORY)
+    logger.trace(s"GeoMesaBatchWriter config: maxMemory set to $memory bytes.")
+    bwc.setMaxMemory(memory)
+
+    val threads = fetchProperty(WRITER_THREADS).map(_.toInt).getOrElse(DEFAULT_THREADS)
+    logger.trace(s"GeoMesaBatchWriter config: maxWriteThreads set to $threads.")
+    bwc.setMaxWriteThreads(threads.toInt)
+
+    fetchProperty(WRITE_TIMEOUT).map { timeout =>
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: should be a foreach

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/AccumuloDataStore.scala [hunk 5]
```diff
@@ -120,9 +121,8 @@ class AccumuloDataStore(val connector: Connector,
   private val visibilityCheckCache = new mutable.HashMap[(String, String), Boolean]()
                                          with mutable.SynchronizedMap[(String, String), Boolean]
 
-  // TODO memory should be configurable
   private val defaultBWConfig =
-    new BatchWriterConfig().setMaxMemory(10000L).setMaxWriteThreads(writeThreads)
+    GeoMesaBatchWriterConfig().setMaxMemory(10000L).setMaxWriteThreads(writeThreads)
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/AccumuloDataStore.scala]: Should we not override the memory here? Makes it so the command line params get ignored...
Overriding in StatWriter and MetaData seems reasonable as those should have smaller memory requirements.

**Comment by elahrvivaz** [comment]
merged as 395eda2bb0bba9f0f8f1f4c714f2d11d7756ea4e

**Pull request closed without merging**